RecoLocalCalo/HGCalRecAlgos/src/HGCalImagingAlgo.cc

```diff
@@ -0,0 +1,516 @@
+      shareEnergy(current_v[i],seeds,fractions);
+
+      // reset and run second pass after vetoing seeds
+      // that result in trivial clusters (less than 2 effective cells)
+      
+
+      for( unsigned isub = 0; isub < fractions.size(); ++isub ) {
+	double effective_hits = 0.0;
+	double energy  = calculateEnergyWithFraction(current_v[i],fractions[isub]);
+	Point position = calculatePositionWithFraction(current_v[i],fractions[isub]);
+
+	//std::cout << "Fractions*Energies: ";
+	for( unsigned ihit = 0; ihit < fractions[isub].size(); ++ihit ) {
+	  const double fraction = fractions[isub][ihit];
+	  if( fraction > 1e-7 ) {
+	    //std::cout << fraction << "*" << current_v[i][ihit].weight << " ";
+	    effective_hits += fraction;
+	    thisCluster.emplace_back(current_v[i][ihit].data.detid,fraction);
+	  }
+	}
+	//std::cout << std::endl;
+	
+	if (verbosity < pINFO)
+	  { 
+	    std::cout << "\t******** NEW CLUSTER (SHARING) ********" << std::endl;
+	    std::cout << "\tEff. No. of cells = " << effective_hits << std::endl;
+	    std::cout << "\t     Energy       = " << energy << std::endl;
+	    std::cout << "\t     Phi          = " << position.phi() << std::endl;
+	    std::cout << "\t     Eta          = " << position.eta() << std::endl;
+	    std::cout << "\t*****************************" << std::endl;
+	  }
+	clusters_v.push_back(reco::BasicCluster(energy, position, caloID, thisCluster, 
+						algoId));
+	thisCluster.clear();
+      }
+    }else{
+      position = calculatePosition(current_v[i]);    
+      std::vector< KDNode >::iterator it;
+      for (it = current_v[i].begin(); it != current_v[i].end(); it++)
+	{
+	  energy += (*it).data.isHalo ? 0. : (*it).data.weight;
+	  thisCluster.emplace_back(std::pair<DetId, float>((*it).data.detid,((*it).data.isHalo?0.:1.)));
+	};
+      if (verbosity < pINFO)
+	{ 
+	  std::cout << "******** NEW CLUSTER (HGCIA) ********" << std::endl;
+	  std::cout << "Index          " << i                   << std::endl;
+	  std::cout << "No. of cells = " << current_v[i].size() << std::endl;
+	  std::cout << "     Energy     = " << energy << std::endl;
+	  std::cout << "     Phi        = " << position.phi() << std::endl;
+	  std::cout << "     Eta        = " << position.eta() << std::endl;
+	  std::cout << "*****************************" << std::endl;
+	}
+      clusters_v.push_back(reco::BasicCluster(energy, position, caloID, thisCluster, 
+					      algoId));
+      thisCluster.clear();
+    }
+  }
+  return clusters_v; 
+}  
+
+math::XYZPoint HGCalImagingAlgo::calculatePosition(std::vector<KDNode> &v){
+  float total_weight = 0.;
+  float x = 0.;
+  float y = 0.;
+  float z = 0.;
+  for (unsigned int i = 0; i < v.size(); i++){
+    if(!v[i].data.isHalo){
+      total_weight += v[i].data.weight;
+      x += v[i].data.x*v[i].data.weight;
+      y += v[i].data.y*v[i].data.weight;
+      z += v[i].data.z*v[i].data.weight;
+    }
+  }
+  
+  return math::XYZPoint( x/total_weight, 
```

Comment:
@lgray: I'm not finding these protections in the code, but maybe they are present. I am concerned about pathological cases of corrupted data or misconfigured software. Maybe an assert would be a reasonable compromise:
```
assert(total_weight != 0.);

```